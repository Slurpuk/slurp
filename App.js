import 'react-native-gesture-handler';
import React, {useEffect, useState} from 'react';
import {NavigationContainer} from '@react-navigation/native';
import HamburgerSlideBarNavigator from './src/navigation/HamburgerSlideBarNavigator';
import SignUpPage from './src/screens/SignUpPage';
import LogInPage from './src/screens/LogInPage';
import WelcomePages from './src/screens/WelcomePages';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import AsyncStorage from '@react-native-async-storage/async-storage';
import firebase from '@react-native-firebase/app';
import firestore from '@react-native-firebase/firestore';
import {Alert} from 'react-native';
import auth from '@react-native-firebase/auth';

export const GlobalContext = React.createContext();
export default function App() {
  const [isFirstTime, setIsFirstTime] = useState(true);
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [currentUser, setCurrentUser] = useState(auth().currentUser);
  const [userRef, setUserRef] = useState(null);
  const [userObj, setUserObj] = useState(null);
  const [shopsData, setShopsData] = useState([]);
  const [isShopIntro, setIsShopIntro] = useState(false);
  const [currShop, setCurrShop] = useState(shopsData[0]);
  const [isFullScreen, setFullScreen] = useState(false);
  const [basketContent, setBasketContent] = useState([]);
  const [basketSize, setBasketSize] = useState(0);
  const [total, setTotal] = useState(0);

  const checkForFirstTime = async () => {
    const result = await AsyncStorage.getItem('isFirstTime');
    //if what we get from the Async is null we are opening the app for the first time
    //if we pressed the sign up button on the last slide we set the 'isFirstTime' to 'no'
    if (result === null) {
      setIsFirstTime(true);
    } //now we can use the isFirstTimeLoad state to choose what to render
  };

  useEffect(() => {
    checkForFirstTime();
  }, []);

  useEffect(() => {
    const subscriber = firebase.auth().onAuthStateChanged(user => {
      if (user) {
        setIsLoggedIn(true);
        setCurrentUser(user);
        setUser();
      } else {
        setIsLoggedIn(false);
        setCurrentUser(null);
      }
    });
    // Unsubscribe from events when no longer in use
    return () => subscriber();
  });

  const enterApp = () => {
    setIsFirstTime(false);
    AsyncStorage.setItem('isFirstTime', 'potatoesInPower');
  };

  function clearBasket() {
    setBasketContent([]);
    setBasketSize(0);
    setTotal(0);
  }

  function newShop({shop, navigation}) {
    clearBasket();
    setCurrShop(shop);
    navigation.navigate('Shop page');
  }

  useEffect(() => {
    const subscriber = firestore()
      .collection('Users')
      .doc(userRef)
      .onSnapshot(documentSnapshot => {
        setUserObj(documentSnapshot.data());
      });

    // Stop listening for updates when no longer required
    return () => subscriber();
  }, [userRef]);

  async function setUser() {
    if (currentUser) {
      await firestore()
        .collection('Users')
        .where('authID', '==', currentUser.uid)
        .get()
        .then(querySnapshot => {
          querySnapshot.forEach(documentSnapshot => {
            setUserRef(documentSnapshot.id);
          });
        });
    }
  }

  function changeShop({shop, navigation}) {
    if (currShop !== shop && basketSize !== 0) {
      Alert.alert(
        'Are you sure ?',
        'Changing shops will clear your basket.',
        [
          {
            text: 'Yes',
            onPress: () => newShop({shop, navigation}),
          },
          {
            text: 'No',
            onPress: () => navigation.navigate('Shop page'),
            style: 'cancel',
          },
        ],
        {cancelable: false},
      );
    } else {
      setCurrShop(shop);
      navigation.navigate('Shop page');
    }
  }

  // Subscribe to the Shops model
  useEffect(() => {
    const subscriber = firestore()
      .collection('CoffeeShop')
      .onSnapshot(querySnapshot => {
        const shops = [];

        querySnapshot.forEach(documentSnapshot => {
          let shopData = {
            ...documentSnapshot.data(),
            key: documentSnapshot.id,
          };

          let coffees = [];
          let drinks = [];
          let snacks = [];
          documentSnapshot.data().ItemsOffered.forEach(itemRef => {
            firestore()
              .doc(itemRef.path)
              .onSnapshot(querySnapshot => {
                let collection = '';
                if (itemRef.path.includes('Coffees')) {
                  collection = coffees;
                } else if (itemRef.path.includes('Drinks')) {
                  collection = drinks;
                } else {
                  collection = snacks;
                }
                collection.push({
                  ...querySnapshot.data(),
                  key: querySnapshot.id,
                });
              });
          });
          shopData.ItemsOffered = {
            Coffees: coffees,
            Drinks: drinks,
            Snacks: snacks,
          };
          shops.push(shopData);
          setShopsData(shops);
          setCurrShop(shops[0]);
        });
      });

    // Unsubscribe from events when no longer in use
    return () => subscriber();
  }, []);

  function isSameItem(it, currIt) {
    if (currIt.hasOwnProperty('Bean') && it.hasOwnProperty('Bean')) {
      let itOptions = '';
      it.options.forEach(option => (itOptions += option.Name));
      let currItOptions = '';
      currIt.options.forEach(option => (currItOptions += option.Name));
      return it.key === currIt.key && itOptions === currItOptions;
    } else {
      return it.key === currIt.key;
    }
  }

  function addToBasket(item) {
    const basket = basketContent;
    const exist = basket.find(x => isSameItem(x, item));
    if (exist) {
      setBasketContent(
        basket.map(x =>
          isSameItem(x, item) ? {...exist, count: exist.count + 1} : x,
        ),
      );
    } else {
      setBasketContent([...basket, {...item, count: 1}]);
    }
    setTotal(total + item.Price);
    setBasketSize(basketSize + 1);
  }

  function removeFromBasket(item) {
    const basket = basketContent;
    const exist = basket.find(x => isSameItem(x, item));
    if (exist.count === 1) {
      setBasketContent(basket.filter(x => x.key !== item.key));
    } else {
      setBasketContent(
        basket.map(x =>
          isSameItem(x, item) ? {...exist, count: exist.count - 1} : x,
        ),
      );
    }
    setTotal(total - item.Price);
    setBasketSize(basketSize - 1);
  }

  const setShopIntro = () => {
    setIsShopIntro(!isShopIntro);
  };

  const Stack = createNativeStackNavigator();
  return (
    <GlobalContext.Provider
      value={{
        enterApp: enterApp,
        user: currentUser, // Returns the authentication object
        currShop: currShop,
        setCurrShop: changeShop,
        isShopIntro: isShopIntro,
        shopsData: shopsData,
        setShopIntro: setShopIntro,
        isFullScreen: isFullScreen,
        setFullScreen: setFullScreen,
        basketContent: basketContent,
        setBasketContent: setBasketContent,
        total: total,
        setTotal: setTotal,
        addToBasket: addToBasket,
        removeFromBasket: removeFromBasket,
        basketSize: basketSize,
        clearBasket: clearBasket,
        currentUser: userObj, // Returns the model object
        userRef: userRef,
      }}
    >
      <NavigationContainer>
        {isLoggedIn ? (
          <HamburgerSlideBarNavigator />
        ) : (
          <Stack.Navigator
            screenOptions={{
              headerShown: false,
            }}
          >
            {isFirstTime ? (
              <Stack.Screen name="Welcome" component={WelcomePages} />
            ) : null}
            <Stack.Screen name="LogIn" component={LogInPage} />
            <Stack.Screen name="SignUp" component={SignUpPage} />
          </Stack.Navigator>
        )}
      </NavigationContainer>
    </GlobalContext.Provider>
  );
}
