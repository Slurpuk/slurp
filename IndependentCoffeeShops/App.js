import React, {useState} from 'react';
import {Pressable} from 'react-native';
import {BottomSheet} from 'react-native-bottomsheet-reanimated';
import type {Node} from 'react';
//import Inputs from 'DesignMap/src/screens/inputs.js'
import {
  View,
  Text,
  TouchableOpacity,
  TextInput,
  StyleSheet,
} from 'react-native';
import PrimaryButton from './PrimaryButton';
import textStyles from './stylesheets/textStyles';
import {authentication} from './firestore/firebase-config';
import firebase from '@react-native-firebase/app';
import auth from '@react-native-firebase/auth';
import firestore from '@react-native-firebase/firestore';
import SignUpPage from "./components/SignUpPage";
import LoginPage from "./components/LoginPage";
import {rgbaColor} from 'react-native-reanimated/src/reanimated2/Colors';

export default function App() {
  const usersCollection = firestore().collection('Users');
  const [isSignedIn, setIsSignedIn] = useState(false);
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [x, setX] = useState(usersCollection);

  const registerUser = () => {
    auth()
      .createUserWithEmailAndPassword(email, password)
      .then(re => {
        console.log(re);
        console.log(x);
      })
      .catch(re => {
        console.log(re);
      })
  }

  return (
    <View style={styles.mainContainer}>
      <TextInput
        placeholder="email"
        value={email}
        onChangeText={text => setEmail(text)}
      />
      <TextInput
        placeholder="password"
        value={password}
        secureTextEntry={true}
        onChangeText={text => setPassword(text)}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  mainContainer: {
    marginTop: 55,
    flex: 1,
    padding: 10,
  },
});
