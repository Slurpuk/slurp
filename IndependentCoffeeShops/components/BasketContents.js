import React, {useState} from 'react';
import {StyleSheet, View, Text, Image, Pressable, FlatList} from 'react-native';

const BasketContents = () => {
  const [total, setTotal] = useState(0);

  const [Items, setItems] = useState([
    {
      key: 1,
      name: 'Latte',
      amount: 0,
      price: '2.40',
      specifications: ['Oat Milk'],
    },
    {
      key: 2,
      name: 'Cappuccino',
      amount: 0,
      price: '2.30',
      specifications: ['Dairy', 'Caramel Syrup'],
    },
    {
      key: 3,
      name: 'Americano',
      amount: 0,
      price: '2.10',
      specifications: [],
    },
    {
      key: 4,
      name: 'Cappuccino',
      amount: 0,
      price: '2.30',
      specifications: ['Dairy', 'Caramel Syrup'],
    },
    {
      key: 5,
      name: 'Cappuccino',
      amount: 0,
      price: '2.30',
      specifications: ['Dairy', 'Caramel Syrup'],
    },
  ]);

  const changeAmount = (item, amount) => {
    item.amount += amount;
    let add_value = Math.round(item.price * amount * 100) / 100;
    console.log(add_value);
    setTotal(total + add_value);
  };

  return (
    <View style={styles.main_container}>
      <View style={styles.basket_content}>
        <Text style={styles.my_order}>My Order</Text>
        <FlatList
          data={Items}
          renderItem={({item}) => (
            <View style={styles.item_container}>
              <View style={styles.item_information}>
                <Text style={styles.item_name}>{item.name}</Text>
                <View style={styles.item_specification_list}>
                  <Text style={styles.item_specification}>Oat Milk</Text>
                </View>
              </View>
              <View style={styles.amount_selection_container}>
                <Text style={styles.amount}>{item.amount}</Text>
                <Pressable onPress={() => changeAmount(item, 1)}>
                  {/*<Image*/}
                  {/*  source={require('../static/downArrow.png')}*/}
                  {/*  style={styles.change_amount_button}*/}
                  {/*/>*/}
                  <Text style={styles.change_amount_button}>+</Text>
                </Pressable>
              </View>
              <Text style={styles.price}>£{(item.price * item.amount).toFixed(2)}</Text>
            </View>
          )}
          style={styles.items_list}
        />

        <View style={styles.order_summary}>
          <Text style={styles.total_text}>TOTAL</Text>
          <Text style={styles.total_amount}>£{total.toFixed(2)}</Text>
        </View>

        <View style={styles.buttons}>
          <Text style={styles.button}>FIRST BEAUTIFUL BUTTON HERE</Text>
          <Text style={styles.button}>SECOND BEAUTIFUL BUTTON HERE</Text>
        </View>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  main_container: {
    backgroundColor: '#E5E5E5',
    display: 'flex',
    height: '100%',
    padding: '5%',
  },
  basket_content: {
    flex: 1,
    display: 'flex',
    height: '100%',
  },
  my_order: {
    fontWeight: '700',
    fontSize: 26,
    color: '#212121',
    paddingBottom: '5%',
    flex: 0.2,
  },
  items_list: {
    display: 'flex',
    flex: 2,
  },
  item_container: {
    borderColor: '#C0C0C0',
    borderStyle: 'solid',
    borderLeftWidth: 0,
    borderRightWidth: 0,
    borderWidth: 1,
    display: 'flex',
    flexDirection: 'row',
    paddingVertical: '5%',
    width: '100%',
    justifyContent: 'space-between',
  },
  item_information: {
    alignSelf: 'flex-start',
    flex: 1,
  },
  item_name: {
    color: '#173C4F',
    fontSize: 21,
    fontWeight: '600',
    paddingBottom: '2%',
    alignSelf: 'flex-start',
  },
  item_specification_list: {},
  item_specification: {
    color: '#717171',
    fontWeight: '300',
    fontSize: 13,
  },
  price: {
    fontWeight: '600',
    fontSize: 17,
    alignSelf: 'flex-start',
    color: '#434343',
  },
  order_summary: {
    display: 'flex',
    flexDirection: 'row',
    paddingVertical: '5%',
    width: '100%',
    justifyContent: 'space-between',
  },
  total_text: {
    fontSize: 21,
    color: '#173C4F',
    fontWeight: '700',
  },
  total_amount: {
    alignSelf: 'flex-end',
    fontSize: 20,
    color: '#173C4F',
    fontWeight: '700',
  },
  amount_selection_container: {
    display: 'flex',
    alignSelf: 'flex-start',
    flexDirection: 'row',
    backgroundColor: '#1B947E',
    borderRadius: 3,
    paddingVertical: '2%',
    paddingHorizontal: '3%',
    flex: 0.15,
    marginEnd: '5%',
  },
  amount: {
    color: '#F1F1F1',
    fontWeight: '600',
    fontSize: 13,
    flex: 1,
  },
  change_amount_button: {
    // resizeMode: 'contain',
    height: 'auto',
    width: 20,
    color: '#FFFFFF',
  },
  buttons: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
  button: {},
});

export default BasketContents;
