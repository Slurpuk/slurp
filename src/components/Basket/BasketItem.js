import {FlatList, Pressable, Text, View} from 'react-native';
import React, {useContext, useState} from 'react';
import {GlobalContext} from '../../../App';
import {BasketItemStyles} from '../../../stylesheets/ShopStyles';

export default function BasketItem({item}) {
  const context = useContext(GlobalContext);
  const [count, setCount] = useState(item.count);
  const [itemTotal, setItemTotal] = useState(item.Price * item.count);

  /**
   * Reduce the basket item's count, ensuring it does not
   * go below 0 (removed from basket).
   */
  function remove() {
    if (count > 0) {
      context.removeFromBasket(item);
      setCount(count - 1);
      setItemTotal(itemTotal - item.Price);
    }
  }

  /**
   * Increase the basket item's count.
   */
  function add() {
    context.addToBasket(item);
    setCount(count + 1);
    setItemTotal(itemTotal + item.Price);
  }

  return (
    <View style={BasketItemStyles.item_container}>
      <View style={BasketItemStyles.item_information}>
        <Text style={BasketItemStyles.item_name}>{item.Name}</Text>
        <FlatList
          data={item.options}
          renderItem={option => (
            <Text style={BasketItemStyles.item_specification}>
              {option.item.Name} {option.item.Type}
            </Text>
          )}
          style={BasketItemStyles.item_specification_list}
          keyExtractor={it => it.key}
        />
      </View>
      <View style={BasketItemStyles.amount_selection_container}>
        <Pressable onPress={() => remove()}>
          <Text
            style={[
              BasketItemStyles.change_amount_button,
              BasketItemStyles.minus,
            ]}>
            -
          </Text>
        </Pressable>
        <Text style={BasketItemStyles.amount}>{count}</Text>
        <Pressable onPress={() => add()}>
          <Text
            style={[
              BasketItemStyles.change_amount_button,
              BasketItemStyles.plus,
            ]}>
            +
          </Text>
        </Pressable>
      </View>
      <Text style={BasketItemStyles.price}>Â£{itemTotal.toFixed(2)}</Text>
    </View>
  );
}
