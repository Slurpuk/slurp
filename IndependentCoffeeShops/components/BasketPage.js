import React from 'react';
import {StyleSheet, Text, View} from 'react-native';
import BasketHeader from './BasketHeader';
import BasketContents from './BasketContents';

const BasketPage = () => {
  return (
    <>
      <SafeAreaView style={styles.safe_header}>

      </SafeAreaView>
      <SafeAreaView style={styles.basket}>
        <BasketHeader style={styles.header} coffeShopName={'ETEN & DRIKEN'} />
        <View style={styles.main_container}>
          <BasketContents/>
        </View>
        <View style={styles.buttons}>
          <Text style={styles.button}>FIRST BEAUTIFUL BUTTON HERE</Text>
          <Text style={styles.button}>SECOND BEAUTIFUL BUTTON HERE</Text>
        </View>
      </SafeAreaView>
  </>
  );
};

const styles = StyleSheet.create({
  main_container: {
    backgroundColor: '#E5E5E5',
    display: 'flex',
    height: '100%',
    padding: '5%',
  },
  basket: {
    display: 'flex',
  },
  header: {
    flex: 1,
  },
  content: {
    flex: 3,
  },
  buttons: {
    flex: 1,
    display: 'flex',
    alignSelf: 'center',
  },
});

export default BasketPage;
