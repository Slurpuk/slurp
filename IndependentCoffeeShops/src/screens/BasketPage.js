import React from 'react';
import {StyleSheet, Text, View, Fragment, SafeAreaView} from 'react-native';
import BasketHeader from '../components/Basket/BasketHeader';
import BasketContents from '../components/Basket/BasketContents';
import PrimaryButton from "../SubComponents/PrimaryButton";

const BasketPage = () => {
  return (
    <>
      <SafeAreaView style={styles.safe_header}>

      </SafeAreaView>
      <SafeAreaView style={styles.basket}>
        <BasketHeader style={styles.header} coffeShopName={'ETEN & DRIKEN'} />
        <View style={styles.main_container}>
          <BasketContents/>
        </View>
        <View style={styles.buttons}>
          <Text style={styles.button}>APPLE/GOOGLE PAY HERE</Text>
          <PrimaryButton text={'Checkout with card'}/>
        </View>
      </SafeAreaView>
  </>
  );
};

const styles = StyleSheet.create({
  basket: {
    flex: 1,
    backgroundColor: '#E5E5E5',
  },
  safe_header: {
    flex: 0,
    backgroundColor: '#046D66',
  },
  header: {
    flex: 1,
  },
  main_container: {
    flex: 11,
    padding: '5%',
  },
  buttons: {
    flex: 1,
    display: 'flex',
    alignSelf: 'center',
    paddingVertical: '10%',
  },
});

export default BasketPage;
