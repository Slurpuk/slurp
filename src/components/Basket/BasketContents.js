import React from 'react';
import {View, Text, FlatList} from 'react-native';
import BasketItem from './BasketItem';
import {basketContentStyles} from './basketStyles';

const BasketContents = ({Items}) => {
  function keyExtract(it) {
    if (it.has_options) {
      let newKey = it.key;
      it.options.forEach(option => (newKey += option.name));
      return newKey;
    } else {
      return it.key;
    }
  }
  return (
    <View style={basketContentStyles.basket_content}>
      <Text style={basketContentStyles.my_order} testID="my-order">My Order</Text>
      <FlatList
        data={Items}
        renderItem={({item}) => <BasketItem item={item} />}
        style={basketContentStyles.items_list}
        keyExtractor={keyExtract}
      />
    </View>
  );
};

export default BasketContents;
