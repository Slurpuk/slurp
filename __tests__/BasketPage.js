import {fireEvent, render, waitFor} from '@testing-library/react-native';
import React from 'react';
import {GlobalContext} from '../App';
import BasketPage from '../src/screens/BasketPage';
import {Alert, Text, View} from "react-native";
import BasketContents from "../src/components/Basket/BasketContents";
import {findAllByText, getByText, queryAllByText, queryByTestId, queryByText} from "@testing-library/react";
import '@testing-library/jest-native/extend-expect';
import {toBeEmpty, toHaveTextContent} from "@testing-library/jest-dom/dist/matchers";
import BasketItem from "../src/components/Basket/BasketItem";

expect.extend({ toBeEmpty, toHaveTextContent });

const globalContextMock = {
  basketContent: null,
  currShop: {Name: 'Eten & Driken', key: 1},
  total: 2.5,
  currentUser: {key: 1},
  clearBasket: null,
    addToBasket: jest.fn(),
    removeFromBasket: jest.fn(),
};

describe('Basket page', function () {
  const spyAlert = jest
      .spyOn(Alert, 'alert')
      //@ts-ignore
      .mockImplementation((title, message, callbackOrButtons) =>
          callbackOrButtons[1].onPress(),
      );

    afterEach(() => spyAlert.mockClear());


  it('Should render correctly', function () {

    const {getByTestId, getByText, queryByTestId, toJSON} = render(
      <GlobalContext.Provider value={globalContextMock}>
        <BasketPage />
      </GlobalContext.Provider>,
    );
    expect(true).toBe(true);
    expect(toJSON()).toMatchSnapshot();
  });

  it('Should render the price rounded to 2 decimal places', function () {


    const {getByText, queryByTestId, toJSON} = render(
        <GlobalContext.Provider value={globalContextMock}>
          <BasketPage />
        </GlobalContext.Provider>,
    );
    expect(getByText('TOTAL')).toBeTruthy();
    expect(getByText('£2.50')).toBeTruthy();

  });

  it("Should render the correct shop name", function (){

    const {getByText, queryByTestId, toJSON} = render(
        <GlobalContext.Provider value={globalContextMock}>
          <BasketPage />
        </GlobalContext.Provider>,
    );
    expect(getByText(`${globalContextMock.currShop.Name}`, {exact: false}))
  })

  it("Should not allow checkout with an empty basket", function (){

    globalContextMock.total = 0;

    const {getByText} = render(
        <GlobalContext.Provider value={globalContextMock}>
          <BasketPage />
        </GlobalContext.Provider>,
    );

    fireEvent(getByText('Checkout'), 'press');
    expect(spyAlert).toHaveBeenCalled();
    expect(spyAlert.mock.calls[0][0]).toBe('Empty basket.');

  })

  it("Should not create an alert with a valid basket", function (){

    globalContextMock.total = 2.50;
    globalContextMock.basketContent = [{key: 1, ItemRef: 'Americano', Quantity: 1, Price: 2.50, Type: 'Coffee', Options: null}];

    const {getByText} = render(
        <GlobalContext.Provider value={globalContextMock}>
          <BasketPage />
        </GlobalContext.Provider>,
    );

    fireEvent(getByText('Checkout'), 'press');
    expect(spyAlert).toBeCalledTimes(0);

  })
});


describe('Basket Content', function (){

  let testItems = [
      {key: 1, ItemRef: 'Americano', Quantity: 1, Price: 2.50, Type: 'Coffee', options: [{Name: 'Dairy', Type: 'Milk', key: 1}], Bean:'Kenyan Single Origin'},
      {key: 1, ItemRef: 'Espresso', Quantity: 2, Price: 1.70, Type: 'Coffee', options: [{Name: 'Oat', Type: 'Milk', key: 1}], Bean:'Ethiopian Single Origin'},
      {key: 1, ItemRef: 'Latte', Quantity: 3, Price: 1.70, Type: 'Coffee', options: [{Name: 'Soy', Type: 'Milk', key: 1}], Bean:'Kenyan Blend'},
      {key: 1, ItemRef: 'Latte', Quantity: 1, Price: 2.40, Type: 'Coffee', options: [{Name: 'Coconut', Type: 'Milk', key: 1}], Bean:'Yucky Nescafe'},
      {key: 1, ItemRef: 'Latte', Quantity: 1, Price: 2.40, Type: 'Coffee', options: [{Name: 'Coconut', Type: 'Milk', key: 1}, {Name: 'Caramel', Type: 'Syrup', key: 2}], Bean:'Yucky Nescafe'},
      {key: 1, ItemRef: 'Croissant', Quantity: 1, Price: 3.50, Type: 'Snack'},
  ]

  it('Should render a number of items equal to the number of unique recipes in basket', function (){
    const {getByText, getAllByText} = render(
          <BasketContents Items={testItems}/>
    );
    expect(getAllByText('+', {exact: false})).toHaveLength(testItems.length);
    expect(getAllByText('-', {exact: false})).toHaveLength(testItems.length);
  });

  it('Should not render any items when the basket is empty', async () => {
      const {queryAllByText, getByText, findAllByText} = render(
          <BasketContents Items={[]}/>
      );
      //It is safe to assume, if there are none of these symbols on the page, no basket items have been rendered.
      expect(await queryAllByText('+')).toHaveLength(0);
      expect(await queryAllByText('-')).toHaveLength(0);
      expect(await queryAllByText('£')).toHaveLength(0);
  });

  it('Should correctly increase the total when the plus icon is pressed', async function (){

      const singularItem = {count: 1, ItemRef: 'Americano', Quantity: 1, Price: 2.50, Type: 'Coffee', options: [{Name: 'Dairy', Type: 'Milk', key: 1}], Bean:'Kenyan Single Origin'}

      const {getByText, queryAllByText} = render(
    
          <GlobalContext.Provider value={globalContextMock}>
            <BasketItem item={singularItem}/>
          </GlobalContext.Provider>
      );

      // Need to add addToBasket in mockGlobal context for this to work
      expect(getByText('1')).toBeTruthy();
      expect(await queryAllByText('2')).toHaveLength(0);

      fireEvent(getByText('+'), 'press');

      expect(await queryAllByText('1')).toHaveLength(0);
      expect(getByText('2')).toBeTruthy();

  });

    it('Should correctly decrease the total when the minus icon is pressed', async function (){

        const singularItem = {count: 2, ItemRef: 'Americano', Quantity: 2, Price: 2.50, Type: 'Coffee', options: [{Name: 'Dairy', Type: 'Milk', key: 1}], Bean:'Kenyan Single Origin'}

        const {getByText, queryAllByText} = render(

            <GlobalContext.Provider value={globalContextMock}>
                <BasketItem item={singularItem}/>
            </GlobalContext.Provider>
        );

        // Need to add addToBasket in mockGlobal context for this to work
        expect(getByText('2')).toBeTruthy();
        expect(await queryAllByText('1')).toHaveLength(0);

        fireEvent(getByText('-'), 'press');

        expect(await queryAllByText('2')).toHaveLength(0);
        expect(getByText('1')).toBeTruthy();

    })

})