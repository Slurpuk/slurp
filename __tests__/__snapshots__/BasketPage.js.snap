// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basket page Should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#E5E5E5",
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#046D66",
        "flexDirection": "row",
        "paddingBottom": "6%",
        "paddingHorizontal": "5%",
        "paddingTop": undefined,
      }
    }
  >
    <View>
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "opacity": 1,
              "underlayColor": "gray",
            },
            Object {},
          ]
        }
        underlayColor="transparent"
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "white",
                "fontSize": 34,
              },
              undefined,
              Object {
                "fontFamily": undefined,
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "marginTop": "2%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#EDEBE7",
            "fontFamily": "JosefinSans-Bold",
            "fontSize": 27,
            "fontStyle": "normal",
            "includeFontPadding": false,
            "letterSpacing": 0.4,
            "marginLeft": 20,
          }
        }
      >
        My Basket - Eten & Driken
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flexShrink": 10,
        "marginHorizontal": "5%",
        "marginVertical": "3%",
      }
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "height": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#212121",
            "fontSize": 26,
            "fontWeight": "700",
            "paddingBottom": "5%",
          }
        }
        testID="my-order"
      >
        My Order
      </Text>
      <RCTScrollView
        data={null}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        style={
          Object {
            "borderColor": "#C0C0C0",
            "borderLeftWidth": 0,
            "borderRightWidth": 0,
            "borderStyle": "solid",
            "borderWidth": 1,
            "display": "flex",
            "flex": 2,
          }
        }
        viewabilityConfigCallbackPairs={Array []}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "maxWidth": "100%",
        "paddingHorizontal": "5%",
        "paddingVertical": "5%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#173C4F",
          "fontSize": 21,
          "fontWeight": "900",
        }
      }
    >
      TOTAL
    </Text>
    <Text
      style={
        Object {
          "color": "#173C4F",
          "fontSize": 20,
          "fontWeight": "900",
        }
      }
    >
      £
      2.50
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "marginBottom": "6%",
        },
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
          "marginVertical": "2%",
        },
      ]
    }
  >
    <View>
      <View
        collapsable={false}
        style={
          Object {
            "transform": Array [
              Object {
                "scale": 1,
              },
            ],
          }
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#087562",
              "borderRadius": 13,
              "display": "flex",
              "flexDirection": "row",
              "height": 43,
              "justifyContent": "center",
              "opacity": 1,
              "width": 682.5,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#EFEFEF",
                "fontFamily": "Poppins-SemiBold",
                "fontSize": 17,
                "letterSpacing": 0.5,
              }
            }
          >
            Checkout
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
